### PR TITLE
feat(dns): Support alt domains

### PR DESCRIPTION
**Pull request by akshayganeshen**
Alternate DNS domains are supplied with the `-alt-domain` flag or `alt_domains` configuration parameter.

The DNS resolver will attempt to parse and match each domain during resolution.

See #4165 for use-case.

### CODE REVIEW

**Comment by freddygv** [comment]
Thanks for the PR @akshayganeshen!

At first glance it seems like one of the new tests is consistently failing in CI: [TestDNS_AltDomains_Overlap](https://travis-ci.org/hashicorp/consul/jobs/543522463)

Is this test passing locally?

**Comment by akshayganeshen** [comment]
Hey @freddygv, that test *does not* pass locally. It's an example of the sort of edge-cases that my implementation does not handle:

```
# alt_domains = [ "dc.consul.", "consul." ]
dig @127.0.0.1 -p 8600 web.service.dc.consul
# is the domain `consul`, and the datacenter `dc`?
# or is the domain `dc.consul`?
```

Without that test, I think it fulfills the desired use-case, but this edge-case should have a well-defined expectation. I believe it can be resolved in two ways:

  1. Simply do not allow these ambiguous `alt_domains`. Raise a validation error in `agent/config/builder.go` when such a scenario is detected. (In my opinion this allows the greatest flexibility to change/define the behaviour later on.)
  2. Prioritize domain matching in some way to break ties between ambiguous domain matches. For example, prefer matching domains that result in queries without a `datacenter`. (That would resolve the scenario I have in the test, but there are other possible edge-cases depending on the prioritization method).

I was hoping to get some feedback before assuming which was the desired solution, or if there's another solution entirely that I'm missing.

File: agent/dns.go [hunk 1]
```diff
@@ -1602,21 +1632,24 @@ func (d *DNSServer) handleRecurse(resp dns.ResponseWriter, req *dns.Msg) {
 // resolveCNAME is used to recursively resolve CNAME records
 func (d *DNSServer) resolveCNAME(cfg *dnsConfig, name string, maxRecursionLevel int) []dns.RR {
 	// If the CNAME record points to a Consul address, resolve it internally
-	// Convert query to lowercase because DNS is case insensitive; d.domain is
-	// already converted
-
-	if strings.HasSuffix(strings.ToLower(name), "."+d.domain) {
```
Comment (i0rek) [inline agent/dns.go]: I think you can check for the `altDomain` suffix here and keep the rest of this function untouched. Looping over the domains shouldn't be necessary.

File: agent/config/config.go [hunk 2]
```diff
@@ -191,6 +191,7 @@ type Config struct {
 	Connect                          Connect                  `json:"connect,omitempty" hcl:"connect" mapstructure:"connect"`
 	DNS                              DNS                      `json:"dns_config,omitempty" hcl:"dns_config" mapstructure:"dns_config"`
 	DNSDomain                        *string                  `json:"domain,omitempty" hcl:"domain" mapstructure:"domain"`
+	DNSAltDomains                    []string                 `json:"alt_domains,omitempty" hcl:"alt_domains" mapstructure:"alt_domains"`
```
[the remark this answers is not in the document]
Comment (akshayganeshen) [reply agent/config/config.go]: True enough. That resolves a lot of complexity added by my changes. Will change it to a single string.

**Comment by akshayganeshen** [comment]
I've updated the implementation to use only a single `alt_domain`. However, I kept the test cases, which includes a failing case, even with the updated `trimDomain`. (Of course, I had to change my input slightly to demonstrate the problem I initially observed.)

Did I miss something there? Or is this as expected?

File: agent/dns.go [hunk 3]
```diff
@@ -529,8 +535,7 @@ func (d *DNSServer) doDispatch(network string, remoteAddr net.Addr, req, resp *d
 	datacenter := d.agent.config.Datacenter
 
 	// Get the QName without the domain suffix
-	qName := strings.ToLower(dns.Fqdn(req.Question[0].Name))
-	qName = strings.TrimSuffix(qName, d.domain)
+	qName := d.trimDomain(dns.Fqdn(req.Question[0].Name))
```
Comment (i0rek) [inline agent/dns.go]: I think you accidentially dropped the `strings.ToLower` here?!

**Comment by clly** [comment]
My personal thought for ambiguous domains is that the primary domain as opposed to the alternative one should be used. 

A second option is to keep a list of the available datacenters to check a potential datacenter. 

I can see a risk where not preferring the previous behavior can cause easily missed issues where a service suddenly fails to discover services that it expected to find in a remote datacenter.

**Comment by i0rek** [comment]
@akshayganeshen @clly I think we found a solution for the ambiguous domains. I suggest we are validating the altDomain in `config/builder.go` like I did in this playground: https://play.golang.org/p/pWJO35B3pV-. That should hopefully resolve the problems. We discovered that not only the datacenter, but also other keywords could potentially cause troubles.

Thanks again, for bearing with us!

**Comment by akshayganeshen** [comment]
@i0rek @clly Thanks for the tips! I'll do the validation in `config/builder.go`, since that's more in line with what I was thinking originally.

(I'll also update the docs appropriately.)

**Comment by akshayganeshen** [comment]
I've updated the implementation and the tests pass!

I'm still updating the documentation though. I'll also add a test case for the invalid config.

File: agent/config/builder.go [hunk 4]
```diff
@@ -1662,3 +1667,16 @@ func isUnixAddr(a net.Addr) bool {
 	_, ok := a.(*net.UnixAddr)
 	return ok
 }
+
+// isValidAltDomain returns true iff the given domain may be used
+// as an alternate domain name.
+func isValidAltDomain(domain, datacenter string) bool {
```
Comment (akshayganeshen) [inline agent/config/builder.go]: Are there any tests to update for this? Or would a test in `dns_test.go` suffice?

Comment (i0rek) [reply agent/config/builder.go]: you could create `agent/config/builder_test.go` and test it separately.

**Comment by akshayganeshen** [comment]
Good catch. Will fix.

**Comment by freddygv** [comment]
@akshayganeshen given that we have a release coming up shortly, I went ahead and made the remaining finishing touches to your PR. 

I hope you don't mind us carrying the PR across the finish line. We appreciate your contribution!

**Comment by akshayganeshen** [comment]
@freddygv @i0rek thank you both! 🙏 

Sorry I couldn't get around to this sooner.

**Pull request merged**